x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java

```diff
@@ -0,0 +1,211 @@
+                        )
+                    );
+                }, listener::onFailure))
+            );
+        }
+    }
+
+    // Package private for testing
+    void getVersionedDocument(Authentication authentication, ActionListener<VersionedDocument> listener) {
+        if (maybeHandleIndexStatusIssue(listener)) {
+            return;
+        }
+
+        final SearchRequest searchRequest = client.prepareSearch(SECURITY_PROFILE_ALIAS)
+            .setQuery(
+                QueryBuilders.boolQuery()
+                    .must(QueryBuilders.termQuery("user.username", authentication.getUser().principal()))
+                    // TODO: this will be replaced by domain lookup and reverse lookup
+                    .must(QueryBuilders.termQuery("user.realm.name", authentication.getSourceRealm().getName()))
+            )
+            .request();
+        try (ThreadContext.StoredContext ignore = client.threadPool().getThreadContext().stashWithOrigin(SECURITY_ORIGIN)) {
+            profileIndex.checkIndexVersionThenExecute(
+                listener::onFailure,
+                () -> executeAsyncWithOrigin(
+                    client,
+                    SECURITY_ORIGIN,
+                    SearchAction.INSTANCE,
+                    searchRequest,
+                    ActionListener.wrap(searchResponse -> {
+                        final SearchHits searchHits = searchResponse.getHits();
+                        final SearchHit[] hits = searchHits.getHits();
+                        if (hits.length < 1) {
+                            logger.debug(
+                                "profile does not exist for username [{}] and realm name [{}]",
+                                authentication.getUser().principal(),
+                                authentication.getSourceRealm().getName()
+                            );
+                            listener.onResponse(null);
+                        } else if (hits.length == 1) {
+                            final SearchHit hit = hits[0];
+                            listener.onResponse(
+                                new VersionedDocument(buildProfileDocument(hit.getSourceRef()), hit.getPrimaryTerm(), hit.getSeqNo())
+                            );
+                        } else {
+                            final ParameterizedMessage errorMessage = new ParameterizedMessage(
+                                "multiple [{}] profiles [{}] found for user [{}]",
+                                hits.length,
+                                Arrays.stream(hits).map(SearchHit::getId).map(this::docIdToUid).sorted().collect(Collectors.joining(",")),
+                                // TODO: include domain information
+                                authentication.getUser().principal()
+                            );
+                            logger.error(errorMessage);
+                            listener.onFailure(new IllegalStateException(errorMessage.getFormattedMessage()));
```

Comment:
It's not quite an illegal state.
I would use an `ElasticsearchException`, we need to discuss how the various APIs handle multiple profiles for the same (username, domain) tuple.

Reply:
I changed it to `ElasticsearchException`. I don't think it makes a real difference though because it's a 500 error on the client side unless we wraps something different inside `ElasticsearchException`.